File: scripts/tasks/sync-openclaw.ts
/**
 * sync-openclaw
 *
 * One-directional sync: OpenClaw workspace → this repo.
 * The OpenClaw workspace is canonical for identity and memory.
 * Code agents should not edit SOUL.md or AGENTS.md in this repo directly;
 * changes flow from the workspace copy outward.
 *
 * Sync map:
 *   ~/.openclaw/workspace/SOUL.md      → <repo>/SOUL.md
 *   ~/.openclaw/workspace/AGENTS.md    → <repo>/AGENTS.md
 *   ~/.openclaw/workspace/IDENTITY.md  → <repo>/openclaw/IDENTITY.md
 *   ~/.openclaw/workspace/USER.md      → <repo>/openclaw/USER.md
 *   ~/.openclaw/workspace/TOOLS.md     → <repo>/openclaw/TOOLS.md
 *   ~/.openclaw/workspace/HEARTBEAT.md → <repo>/openclaw/HEARTBEAT.md
 *   ~/.openclaw/workspace/memory/      → <repo>/openclaw/memory/
 *   ~/.openclaw/cron/jobs.json         → <repo>/openclaw/cron-jobs.json
 *   ~/.openclaw/exec-approvals.json    → <repo>/openclaw/exec-approvals.json
 *
 * Secrets (openclaw.json, credentials/, tokens) are NOT synced here.
 * Use `bun run scry:setup:config_backup` for encrypted secret backups.
 *
 * Usage:
 *   bun run scry:sync:openclaw               # sync only
 *   bun run scry:sync:openclaw -- --commit   # sync + git commit + push
 */

import {
  copyFileSync,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  rmSync,
  statSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { logStep } from "../common";

const OPENCLAW_WORKSPACE =
  process.env.OPENCLAW_WORKSPACE ??
  join(process.env.HOME ?? homedir(), ".openclaw", "workspace");

const OPENCLAW_ROOT =
  process.env.OPENCLAW_ROOT ?? join(process.env.HOME ?? homedir(), ".openclaw");

const REPO_ROOT = resolve(".");

/** Files synced to repo root (identity docs shared with code agents). */
const ROOT_FILES = ["SOUL.md", "AGENTS.md"] as const;

/** Files synced to openclaw/ directory. */
const OPENCLAW_FILES = [
  "IDENTITY.md",
  "USER.md",
  "TOOLS.md",
  "HEARTBEAT.md",
] as const;

/** Extra files from ~/.openclaw/ (not workspace) synced to openclaw/. */
const EXTRA_FILES: ReadonlyArray<{ src: string; dest: string }> = [
  { src: "cron/jobs.json", dest: "cron-jobs.json" },
  { src: "exec-approvals.json", dest: "exec-approvals.json" },
];

type SyncResult = {
  copied: number;
  skipped: number;
  deleted: number;
  errors: string[];
};

function filesMatch(a: string, b: string): boolean {
  if (!existsSync(a) || !existsSync(b)) return false;
  const aStat = statSync(a);
  const bStat = statSync(b);
  if (aStat.size !== bStat.size) return false;
  const aContent = readFileSync(a);
  const bContent = readFileSync(b);
  return aContent.equals(bContent);
}

function ensureDir(path: string): void {
  mkdirSync(path, { recursive: true });
}

function syncFile(src: string, dest: string, result: SyncResult): void {
  if (!existsSync(src)) return;

  if (filesMatch(src, dest)) {
    result.skipped++;
    return;
  }

  try {
    ensureDir(dirname(dest));
    copyFileSync(src, dest);
    console.log(`  [COPY] ${src} → ${dest}`);
    result.copied++;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`  [ERROR] ${src}: ${msg}`);
    result.errors.push(msg);
  }
}

function syncMemory(result: SyncResult): void {
  const srcDir = join(OPENCLAW_WORKSPACE, "memory");
  const destDir = join(REPO_ROOT, "openclaw", "memory");

  if (!existsSync(srcDir)) {
    console.log("  [SKIP] no memory directory in workspace");
    return;
  }

  ensureDir(destDir);

  const srcFiles = new Set(
    readdirSync(srcDir).filter((f) => f.endsWith(".md")),
  );
  const destFiles = new Set(
    existsSync(destDir)
      ? readdirSync(destDir).filter((f) => f.endsWith(".md"))
      : [],
  );

  // Copy new/updated memory files
  for (const file of srcFiles) {
    syncFile(join(srcDir, file), join(destDir, file), result);
  }

  // Remove memory files that no longer exist in workspace
  for (const file of destFiles) {
    if (!srcFiles.has(file)) {
      const destPath = join(destDir, file);
      try {
        rmSync(destPath);
        console.log(`  [DELETE] ${destPath}`);
        result.deleted++;
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        console.error(`  [ERROR] delete ${destPath}: ${msg}`);
        result.errors.push(msg);
      }
    }
  }
}

export function syncOpenclaw(): void {
  const commit = Bun.argv.includes("--commit");

  logStep("Syncing OpenClaw workspace → repo");
  console.log(`  workspace: ${OPENCLAW_WORKSPACE}`);
  console.log(`  repo:      ${REPO_ROOT}`);

  if (!existsSync(OPENCLAW_WORKSPACE)) {
    throw new Error(`OpenClaw workspace not found: ${OPENCLAW_WORKSPACE}`);
  }

  const result: SyncResult = { copied: 0, skipped: 0, deleted: 0, errors: [] };

  // Sync identity docs to repo root
  logStep("Identity docs → repo root");
  for (const file of ROOT_FILES) {
    syncFile(join(OPENCLAW_WORKSPACE, file), join(REPO_ROOT, file), result);
  }

  // Sync workspace files to openclaw/
  logStep("Workspace files → openclaw/");
  for (const file of OPENCLAW_FILES) {
    syncFile(
      join(OPENCLAW_WORKSPACE, file),
      join(REPO_ROOT, "openclaw", file),
      result,
    );
  }

  // Sync extra files from ~/.openclaw/
  logStep("OpenClaw config files → openclaw/");
  for (const { src, dest } of EXTRA_FILES) {
    syncFile(
      join(OPENCLAW_ROOT, src),
      join(REPO_ROOT, "openclaw", dest),
      result,
    );
  }

  // Sync memory
  logStep("Memory → openclaw/memory/");
  syncMemory(result);

  // Summary
  logStep("Sync complete");
  console.log(`  copied: ${result.copied}`);
  console.log(`  skipped: ${result.skipped} (unchanged)`);
  console.log(`  deleted: ${result.deleted}`);
  if (result.errors.length > 0) {
    console.log(`  errors: ${result.errors.length}`);
  }

  if (!commit) {
    if (result.copied > 0 || result.deleted > 0) {
      console.log("\n  pass --commit to auto-commit and push");
    }
    return;
  }

  if (result.copied === 0 && result.deleted === 0) {
    console.log("\n  nothing changed; skipping commit");
    return;
  }

  logStep("Committing and pushing");
  const date = new Date().toISOString().split("T")[0];

  const gitAdd = Bun.spawnSync({
    cmd: ["git", "-C", REPO_ROOT, "add", "-A"],
    stdout: "pipe",
    stderr: "pipe",
  });
  if (gitAdd.exitCode !== 0) {
    throw new Error("git add failed");
  }

  const gitCommit = Bun.spawnSync({
    cmd: [
      "git",
      "-C",
      REPO_ROOT,
      "commit",
      "-m",
      `sync openclaw workspace ${date}`,
    ],
    stdout: "pipe",
    stderr: "pipe",
  });
  if (gitCommit.exitCode !== 0) {
    const stderr = Buffer.from(gitCommit.stderr).toString("utf8").trim();
    if (stderr.includes("nothing to commit")) {
      console.log("  nothing to commit (git tree clean)");
      return;
    }
    throw new Error(`git commit failed: ${stderr}`);
  }

  const gitPush = Bun.spawnSync({
    cmd: ["git", "-C", REPO_ROOT, "push", "origin", "main"],
    stdout: "pipe",
    stderr: "pipe",
  });

  const pushStderr = Buffer.from(gitPush.stderr).toString("utf8").trim();
  if (gitPush.exitCode !== 0) {
    throw new Error(`git push failed: ${pushStderr}`);
  }

  if (pushStderr.length > 0) {
    console.log(`  ${pushStderr}`);
  }
  console.log("  committed and pushed");
}

File: scripts/tasks/sync_openclaw.py
"""One-directional sync: OpenClaw workspace → this repo.

The OpenClaw workspace is canonical for identity and memory.
Specialist workspaces are mirrored for doc/back-up coverage.

Sync map:
  ~/.openclaw/workspace/SOUL.md            → <repo>/SOUL.md
  ~/.openclaw/workspace/AGENTS.md          → <repo>/AGENTS.md
  ~/.openclaw/workspace/*.md               → <repo>/openclaw/*.md
  ~/.openclaw/workspace/memory/**/*.md     → <repo>/openclaw/memory/
  ~/.openclaw/workspace/docs/**/*.md       → <repo>/openclaw/docs/
  ~/.openclaw/workspace/prompts/**/*.md    → <repo>/openclaw/prompts/
  ~/.openclaw/workspace-*/**/*.md          → <repo>/openclaw/specialists/<agentId>/
  ~/.openclaw/workspace-*/scripts/         → <repo>/openclaw/specialists/<agentId>/scripts/
  ~/.openclaw/workspace-*/templates/       → <repo>/openclaw/specialists/<agentId>/templates/
  ~/.openclaw/workspace-*/hooks/           → <repo>/openclaw/specialists/<agentId>/hooks/
  ~/.openclaw/cron/jobs.json               → <repo>/openclaw/cron-jobs.json
  ~/.openclaw/exec-approvals.json          → <repo>/openclaw/exec-approvals.json

Secrets (openclaw.json, credentials/, tokens) are NOT synced here.

Usage:
  uv run python -m scripts sync:openclaw               # sync only
  uv run python -m scripts sync:openclaw --commit      # sync + git commit + push
"""

from __future__ import annotations

import shutil
import sys
from datetime import UTC, datetime
from pathlib import Path

from scripts.common import ensure_dir, log_step, run_or_throw

OPENCLAW_WORKSPACE = Path(
    __import__("os").environ.get(
        "OPENCLAW_WORKSPACE", Path.home() / ".openclaw" / "workspace"
    )
)
OPENCLAW_ROOT = Path(
    __import__("os").environ.get("OPENCLAW_ROOT", Path.home() / ".openclaw")
)
REPO_ROOT = Path.cwd().resolve()

ROOT_FILES = ["SOUL.md", "AGENTS.md"]
MAIN_MARKDOWN_DIRS = ["memory", "docs", "prompts", "templates"]
SPECIALIST_EXCLUDE = {"main", "claude", "codex"}
SPECIALIST_MARKDOWN_DIRS = ["memory", "docs", "prompts"]
SPECIALIST_DIRS = ["scripts", "templates", "hooks", "coordination"]
EXTRA_FILES = [
    ("cron/jobs.json", "cron-jobs.json"),
    ("exec-approvals.json", "exec-approvals.json"),
]


class SyncResult:
    def __init__(self) -> None:
        self.copied = 0
        self.skipped = 0
        self.deleted = 0
        self.errors: list[str] = []


def _record_error(result: SyncResult, message: str, exc: Exception) -> None:
    detail = f"{message}: {exc}"
    print(f"  [ERROR] {detail}", file=sys.stderr)
    result.errors.append(detail)


def _files_match(a: Path, b: Path) -> bool:
    if not a.exists() or not b.exists():
        return False
    if a.stat().st_size != b.stat().st_size:
        return False
    return a.read_bytes() == b.read_bytes()


def _sync_file(src: Path, dest: Path, result: SyncResult) -> None:
    if not src.exists():
        return

    if _files_match(src, dest):
        result.skipped += 1
        return

    try:
        ensure_dir(dest.parent)
        shutil.copy2(str(src), str(dest))
        print(f"  [COPY] {src} → {dest}")
        result.copied += 1
    except Exception as exc:
        _record_error(result, f"copy {src} -> {dest}", exc)


def _sync_markdown_dir(src_dir: Path, dest_dir: Path, result: SyncResult) -> None:
    if not src_dir.exists():
        return

    ensure_dir(dest_dir)

    src_files = {f.name for f in src_dir.iterdir() if f.is_file() and f.suffix == ".md"}
    dest_files = (
        {f.name for f in dest_dir.iterdir() if f.is_file() and f.suffix == ".md"}
        if dest_dir.exists()
        else set()
    )

    for file in src_files:
        _sync_file(src_dir / file, dest_dir / file, result)

    for file in dest_files:
        if file not in src_files:
            dest_path = dest_dir / file
            try:
                dest_path.unlink()
                print(f"  [DELETE] {dest_path}")
                result.deleted += 1
            except Exception as exc:
                _record_error(result, f"delete {dest_path}", exc)


def _sync_root_markdown_files(
    src_dir: Path,
    dest_dir: Path,
    result: SyncResult,
    *,
    exclude: set[str] | None = None,
) -> None:
    ensure_dir(dest_dir)
    exclude = exclude or set()

    src_files = {
        f.name
        for f in src_dir.iterdir()
        if f.is_file() and f.suffix == ".md" and f.name not in exclude
    }
    dest_files = {
        f.name
        for f in dest_dir.iterdir()
        if f.is_file() and f.suffix == ".md" and f.name not in exclude
    }

    for file in sorted(src_files):
        _sync_file(src_dir / file, dest_dir / file, result)

    for file in sorted(dest_files - src_files):
        dest_path = dest_dir / file
        try:
            dest_path.unlink()
            print(f"  [DELETE] {dest_path}")
            result.deleted += 1
        except Exception as exc:
            _record_error(result, f"delete {dest_path}", exc)


def _sync_markdown_tree(src_dir: Path, dest_dir: Path, result: SyncResult) -> None:
    if not src_dir.exists():
        return

    ensure_dir(dest_dir)

    src_files = {
        p.relative_to(src_dir)
        for p in src_dir.rglob("*.md")
        if p.is_file()
    }
    dest_files = {
        p.relative_to(dest_dir)
        for p in dest_dir.rglob("*.md")
        if p.is_file()
    } if dest_dir.exists() else set()

    for rel in sorted(src_files):
        _sync_file(src_dir / rel, dest_dir / rel, result)

    for rel in sorted(dest_files - src_files, reverse=True):
        dest = dest_dir / rel
        try:
            dest.unlink()
            print(f"  [DELETE] {dest}")
            result.deleted += 1
        except Exception as exc:
            _record_error(result, f"delete {dest}", exc)

    for path in sorted(dest_dir.rglob("*"), reverse=True):
        if path.is_dir():
            try:
                next(path.iterdir())
            except StopIteration:
                path.rmdir()


def _sync_tree(src_dir: Path, dest_dir: Path, result: SyncResult) -> None:
    if not src_dir.exists():
        return

    ensure_dir(dest_dir)

    src_entries = {p.relative_to(src_dir) for p in src_dir.rglob("*")}
    dest_entries = {p.relative_to(dest_dir) for p in dest_dir.rglob("*")} if dest_dir.exists() else set()

    for rel in sorted(src_entries):
        src = src_dir / rel
        dest = dest_dir / rel
        if src.is_dir():
            ensure_dir(dest)
            continue
        _sync_file(src, dest, result)

    for rel in sorted(dest_entries, reverse=True):
        if rel in src_entries:
            continue
        dest = dest_dir / rel
        try:
            if dest.is_dir():
                shutil.rmtree(dest)
            else:
                dest.unlink()
            print(f"  [DELETE] {dest}")
            result.deleted += 1
        except Exception as exc:
            print(f"  [ERROR] delete {dest}: {exc}", file=sys.stderr)
            result.errors.append(str(exc))


def _iter_specialist_ids() -> list[str]:
    if not OPENCLAW_ROOT.exists():
        return []
    return sorted(
        d.name.removeprefix("workspace-")
        for d in OPENCLAW_ROOT.iterdir()
        if d.is_dir()
        and d.name.startswith("workspace-")
        and d.name.removeprefix("workspace-") not in SPECIALIST_EXCLUDE
    )


def _sync_specialists(result: SyncResult) -> None:
    dest_root = REPO_ROOT / "openclaw" / "specialists"
    ensure_dir(dest_root)

    specialist_ids = _iter_specialist_ids()
    seen = set(specialist_ids)

    for agent_id in specialist_ids:
        src_ws = OPENCLAW_ROOT / f"workspace-{agent_id}"
        dest_ws = dest_root / agent_id
        ensure_dir(dest_ws)

        _sync_root_markdown_files(src_ws, dest_ws, result)

        for dirname in SPECIALIST_MARKDOWN_DIRS:
            src_dir = src_ws / dirname
            dest_dir = dest_ws / dirname
            if src_dir.exists():
                _sync_markdown_tree(src_dir, dest_dir, result)
            elif dest_dir.exists():
                try:
                    shutil.rmtree(dest_dir)
                    print(f"  [DELETE] {dest_dir}")
                    result.deleted += 1
                except Exception as exc:
                    _record_error(result, f"delete {dest_dir}", exc)

        for dirname in SPECIALIST_DIRS:
            src_dir = src_ws / dirname
            dest_dir = dest_ws / dirname
            if src_dir.exists():
                _sync_tree(src_dir, dest_dir, result)
            elif dest_dir.exists():
                try:
                    shutil.rmtree(dest_dir)
                    print(f"  [DELETE] {dest_dir}")
                    result.deleted += 1
                except Exception as exc:
                    _record_error(result, f"delete {dest_dir}", exc)

    existing_dirs = {d.name for d in dest_root.iterdir() if d.is_dir()} if dest_root.exists() else set()
    for agent_id in existing_dirs - seen:
        stale = dest_root / agent_id
        try:
            shutil.rmtree(stale)
            print(f"  [DELETE] {stale}")
            result.deleted += 1
        except Exception as exc:
            _record_error(result, f"delete {stale}", exc)


def sync_openclaw() -> None:
    commit = "--commit" in sys.argv

    log_step("Syncing OpenClaw workspace → repo")
    print(f"  workspace: {OPENCLAW_WORKSPACE}")
    print(f"  repo:      {REPO_ROOT}")

    if not OPENCLAW_WORKSPACE.exists():
        raise RuntimeError(f"OpenClaw workspace not found: {OPENCLAW_WORKSPACE}")

    result = SyncResult()

    log_step("Identity docs → repo root")
    for file in ROOT_FILES:
        _sync_file(OPENCLAW_WORKSPACE / file, REPO_ROOT / file, result)

    log_step("Workspace markdown → openclaw/")
    _sync_root_markdown_files(OPENCLAW_WORKSPACE, REPO_ROOT / "openclaw", result)

    log_step("Workspace markdown dirs → openclaw/")
    for dirname in MAIN_MARKDOWN_DIRS:
        src_dir = OPENCLAW_WORKSPACE / dirname
        dest_dir = REPO_ROOT / "openclaw" / dirname
        if src_dir.exists():
            _sync_markdown_tree(src_dir, dest_dir, result)
        elif dest_dir.exists():
            try:
                shutil.rmtree(dest_dir)
                print(f"  [DELETE] {dest_dir}")
                result.deleted += 1
            except Exception as exc:
                _record_error(result, f"delete {dest_dir}", exc)

    log_step("OpenClaw config files → openclaw/")
    for src, dest in EXTRA_FILES:
        _sync_file(OPENCLAW_ROOT / src, REPO_ROOT / "openclaw" / dest, result)

    log_step("Specialist workspace mirrors → openclaw/specialists/")
    _sync_specialists(result)

    log_step("Sync complete")
    print(f"  copied: {result.copied}")
    print(f"  skipped: {result.skipped} (unchanged)")
    print(f"  deleted: {result.deleted}")
    if result.errors:
        print(f"  errors: {len(result.errors)}")
        for error in result.errors:
            print(f"  - {error}")

    if result.errors:
        raise RuntimeError(
            f"OpenClaw sync aborted after {len(result.errors)} filesystem error(s); nothing was committed."
        )

    if not commit:
        if result.copied > 0 or result.deleted > 0:
            print("\n  pass --commit to auto-commit and push")
        return

    if result.copied == 0 and result.deleted == 0:
        print("\n  nothing changed; skipping commit")
        return

    log_step("Committing and pushing")
    date = datetime.now(UTC).strftime("%Y-%m-%d")

    # Only stage the files we actually synced — never `git add -A` which
    # could sweep in unrelated local edits.
    sync_paths = [
        *(str(REPO_ROOT / f) for f in ROOT_FILES),
        str(REPO_ROOT / "openclaw"),
        *(str(REPO_ROOT / "openclaw" / dest) for _, dest in EXTRA_FILES),
        str(REPO_ROOT / "openclaw" / "specialists"),
    ]
    run_or_throw(["git", "add", "--", *sync_paths], cwd=str(REPO_ROOT))

    try:
        run_or_throw(
            ["git", "commit", "-m", f"sync openclaw workspace {date}"],
            cwd=str(REPO_ROOT),
        )
    except RuntimeError as exc:
        if "nothing to commit" in str(exc):
            print("  nothing to commit (git tree clean)")
            return
        raise

    run_or_throw(["git", "push", "origin", "main"], cwd=str(REPO_ROOT))
    print("  committed and pushed")


if __name__ == "__main__":
    sync_openclaw()
